src/orbits/keplerian.jl

```diff
@@ -0,0 +1,532 @@
+            ))
+            omega = nt.omega
+            sin_omega, cos_omega = sincos(nt.omega)
+        elseif all( (!isnothing).((nt.cos_omega, nt.sin_omega)) )
+            cos_omega, sin_omega = nt.cos_omega, nt.sin_omega
+            omega = atan(sin_omega, cos_omega)
+        else
+            throw(ArgumentError("`ω` must also be provided if `ecc` specified"))
+        end
+        E_0 = 2.0 * atan(√(1.0 - ecc) * cos_omega, √(1.0 + ecc) * (1.0 + sin_omega))
+        M_0 = E_0 - ecc * sin(E_0)
+
+        incl_factor_inv  = (1.0 - ecc^2) / (1.0 + ecc * sin_omega)
+    end
+
+    # Jacobian for cos(i) -> b
+    dcosi_db = R_star / a * (1.0 / incl_factor_inv)
+
+    if !isnothing(nt.b)
+        any( (!isnothing).((nt.incl, duration)) ) && throw(ArgumentError(
+            "Only `incl`, `b`, or `duration` can be given"
+        ))
+        b = nt.b
+        cos_incl = dcosi_db * b
+        incl = acos(cos_incl)
+        sin_incl = sin(incl)
+        duration = nt.duration
+    elseif !isnothing(nt.incl)
+        !isnothing(nt.duration) && throw(ArgumentError(
+            "Only `incl`, `b`, or `duration` can be given"
+        ))
+        incl = nt.incl
+        sin_incl, cos_incl = sincos(incl)
+        b = cos_incl / dcosi_db
+        duration = nt.duration
+    elseif !isnothing(nt.duration)
+        duration = nt.duration
+        c = sin(π * duration / (incl_factor_inv) / period)
+        c_sq = c^2
+        ecc_sin_omega = ecc*sin_omega
+        aor = a_planet / R_star
+        b = √(
+            (aor^2 * c_sq - 1.0) /
+            (
+                c_sq * ecc_sin_omega^2 +
+                2.0*c_sq*ecc_sin_omega +
+                c_sq - ecc^4 + 2.0*ecc^2 - 1.0
+            )
+        ) * (1.0 - ecc^2)
+        cos_incl = dcosi_db * b
+        incl = acos(cos_incl)
+        sin_incl = sin(incl)
+    else
+        incl = 0.5*π
+        cos_incl = 0.0
+        sin_incl = 1.0
+        b = 0.0
+        duration = nt.duration
+    end
+
+    # Compute remaining system parameters
+    !(isnothing(nt.t_0) ⊻ isnothing(nt.t_p)) && throw(
+        ArgumentError("Either t₀ or tₚ must be specified")
+    )
+    if isnothing(nt.t_0)
+        t_p = nt.t_p
+        t_0 = t_p + M_0 / n
+    else
+        t_0 = nt.t_0
+        t_p = t_0 - M_0 / n
+    end
+
+    t_ref = t_p - t_0
+
+    # Normalize inputs
+    (
+        period, t_0, t_p, t_ref, duration,
+        a, a_planet, a_star, R_planet, R_star,
+        rho_planet, rho_star,
+        RpRs, aR_star, b, ecc, M_0, cos_incl, sin_incl, cos_omega, sin_omega, cos_Omega, sin_Omega,
+        incl, omega, Omega,
+        M_planet, M_star,
+    ) = normalize_inputs(
+        period, t_0, t_p, t_ref, duration,
+        a, a_planet, a_star, R_planet, R_star,
+        rho_planet, rho_star,
+        RpRs, aR_star, b, ecc, M_0, cos_incl, sin_incl, cos_omega, sin_omega, cos_Omega, sin_Omega,
+        incl, omega, Omega,
+        M_planet, M_star,
+        no_units,
+    )
```

Comment:
I'm not a huge fan of this, I think it is easier for us devs, and ultimately puts more control in the user's hands, if we let the types and stuff interact with themselves when possible.

I especially want to avoid forcing users to use units when they don't have to. For example, if they only use units for timing info, they shouldn't have to provide any units for e.g. `a`. However, currently, things are locked in basically by whether or not `period` and `a` have units.

Looking further into this example, let's say I give "period" as a Unitful quantity, what needs to be calculated from it? The angular frequency `n` depends on it, and later on there is a calculation that looks like `duration / period` basically. If a user doesn't provide `duration` as a quantity the code will error- we have no obligation to assume the duration units for the user. Furthermore, since this error already tells the user what happened, we don't really need to add special case-handling for it either. Just let Julia do it for you.

Does this make sense? Could you try minimizing the amount you manually change the inputs to accomplish this?

Reply:
> For example, if they only use units for timing info, they shouldn't have to provide any units for e.g. `a`.

I don't think I am clear on what you mean here. Do you want the units for `a` to default to unitful, nominal units in this case? For example, `KeplerianOrbit(period=1.0u"d", t_0=0.0u"d", a=7.0)` should assume `a = 7.0u"Rsun"`?

Reply:
Not quite-
your example will use the nominal units for `a`, which are `Rsun` which is `1`. What I'm saying we need to change from the current implementation, is that  `a` shouldn't become a `Unitful.Quantity` unless the user inputs that.

Similarly, I could do something like
```julia
KeplerianOrbit(t0=0, period=5, duration=0.1, a=1u"AU", Rsun=2.2u"Rsun")
```
where I'm not using units for the time (which will assume the nominal units, which in this case is arbitrary but is assumed to be consistent), but I am using units for some of orbital parameters. I would not want this example to automatically add units to the `t0`, `period` or `duration` nor would I want it to strip the units from `a` or `Rsun`.

Reply:
I think `normalize_inputs` could definitely be revisited. Currently it follows a similar pattern to what is used for `SimpleOrbit`:
https://github.com/JuliaAstro/Transits.jl/blob/d7ee355dc719f38f98d9a262629653645ebf146e/src/orbits/simple.jl#L27-L28

> your example will use the nominal units for a, which are Rsun which is 1

So this example should error as expected and not fall back to any implied units I think, so I am not quite clear on what you mean here:
```julia
> KeplerianOrbit(period=1.0u"d", t_0=0.0u"d", a=7.0)
ERROR: DimensionError: 4.602361709975639 d^-2 and 0.0 are not dimensionally compatible
```

> is that a shouldn't become a Unitful.Quantity unless thats what the user inputted.

Similarly, it's my understanding that `a` will not be converted unless we explicitly have the units specified, i.e., `KeplerianOrbit(period=1.0u"d", t_0=0.0u"d", a=7.0u"Rsun")` and `KeplerianOrbit(period=1.0, t_0=0.0, a=7.0)` should return roughly the same thing.

>  I would not want this example to automatically add units to the t0, period or duration nor would I want it to strip the units from a or Rsun

Agreed, and I don't think it does? It fails when the user tries to mix units/nominal unitless inputs.

> For example, if they only use units for timing info, they shouldn't have to provide any units for e.g. a. However, currently, things are locked in basically by whether or not period and a have units.

I think this could definitely be handled more elegantly, but would allowing what your asking for introduce the problems mentioned above? Sorry if I am misunderstanding the point you are making here!